First PR for WholeGraph 23.06 refactor and add RAPIDS CI scripts

dongxuy04 opened the pull request:
Refactor for WholeGraph, some updates:
- More WholeMemory types.
- WholeMemory Embedding.
- WholeMemory Embedding Cache support.
- WholeMemory Embedding Sparse Optimizer.
- DLPack export for WholeMemory.
- Torch C10 API removed.
- Packaging for conda.
- RAPIDS CI scripts.

BradReesWork commented:
@dongxuy04 you need to add a conda environment yml file under  "conda/environments"

dongxuy04 commented:
> @dongxuy04 you need to add a conda environment yml file under "conda/environments"

@BradReesWork added conda environment yml files generated by rapids-dependency-file-generator.

seunghwak commented:
Does WholeGraph provide C-API? or C++-API? Most public headers under the `cpp/include` directory has either `.h` or `.cuh` extensions and should I assume this is to support C users?

dongxuy04 commented:
> Does WholeGraph provide C-API? or C++-API? All the public headers under the `cpp/include` directory has either `.h` or `.cuh` extensions and should I assume this is to support C users?

Yes, for host API part, `.h` files provide C-API instead of C++-API, the initial consideration is not to introduce ABI issues. For device headers, `.cuh` aims to provide developing interface on WholeMemory, so it expose C++ device API to access WholeMemory.

BradReesWork commented:
@dongxuy04 can you rename **_build_components.sh_** to just **_build.sh_** to match other repos?

dongxuy04 commented:
> @dongxuy04 can you rename **_build_components.sh_** to just **_build.sh_** to match other repos?

@BradReesWork I have renamed `build_components.sh` to `build.sh`, and updated scripts called this script.

MatthiasKohl commented:
@BradReesWork this PR has ~50K touched lines, is there something in particular you want me to review?

stadlmax commented on pylibwholegraph/pylibwholegraph/torch/graph_ops.py:
For another time: what is our strategy regarding GNN ops? @teju85 @BradReesWork 
cugraph-ops always has been told to stay closed source. This is mostly duplicating the work we have in cugraph-ops. Further functionalities are duplicating stuff we have in cugraph-pyg and cugraph-dgl. 
As discussed already two times offline, I thought the whole point of this refactoring is to make WholeGraph a storage-only solution.

MatthiasKohl commented:
For every file and kernel under `cpp/src/graph_ops`, can we get a description of:
1. what is the goal? what model this operation would be used for?
2. was cugraph-ops considered? if yes, why was it not possible to use cugraph-ops/what is missing from cugraph-ops?

MatthiasKohl commented:
tagging @BradReesWork since you asked for scope of duplication of operations and @dongxuy04 for visibility for the above comment

teju85 commented:
I understand that this PR only focuses on refactor of WG into a more RAPIDS-like workflow. But I too still want to mention the cugraph-ops integration into WG and urge to put this on our immediate list of things todo for WG.

I agree with both Max's/Matthias' requests. cugraphops should already provide all (if not most) of those operations. So, it's not worth the duplication. I have filed issue #17 for this.

dongxuy04 commented:
> For every file and kernel under `cpp/src/graph_ops`, can we get a description of:
> 
> 1. what is the goal? what model this operation would be used for?
> 2. was cugraph-ops considered? if yes, why was it not possible to use cugraph-ops/what is missing from cugraph-ops?

These files are legacy implementations left over from previous versions of WholeGraph code. Agreed  that we need to remove these duplication and actually this is in processing. The goal might be to get most of these files replaced by cuGraph ops.

Here are the goal of the files under `cpp/src/graph_ops`.
- **append_unique***: AppendUnique op, it is an optimizing OP for ids between sampling of two layers. The action is append sampled neighbor nodes to center nodes, and the do unique while keep center nodes unchanged. May need to keep as it is not related to cuGraphOps.
- **csr_add_self_loop***: add self loop to sampled csr sub graph. Not sure if there is this kind of ops. If yes, we may remove that.
- **edge_weight_softmax***, **gspmm_csr_weighted***, **spadd_gat_csr***: ops related to GAT model, currently GAT module from cuGraphOps is integrating, will be removed after confirmed.
- **spmm_csr_no_weight***: Ops related to GraphSAGE op and GCN op, currently GraphSAGE module from cuGraphOps is integrated, will be removed later.

dongxuy04 commented:
@MatthiasKohl @stadlmax @teju85 @BradReesWork Agreed that it doesn't worth duplication. I have verified that similar ops in cuGraphOps works and removed the duplicated ops under `cpp/src/graph_ops`.

BradReesWork commented on build.sh:
```diff
@@ -0,0 +1,199 @@
+#!/bin/bash
+
```
./build.sh pylibwholegraph
Building for *ALL* supported GPU architectures...
**env: ‘setup.py’: No such file or directory**

BradReesWork commented:
the "cmake" folder should be under the cpp directory

BradReesWork commented:
the top level folder for python code should be called "python" and then all the moudules under that

seunghwak commented on cpp/src/logger.hpp:
You may consider using raft's logging mechanism (e.g. https://github.com/rapidsai/raft/blob/branch-23.06/cpp/include/raft/core/logger-macros.hpp).

dongxuy04 commented:
> the "cmake" folder should be under the cpp directory

dongxuy04 commented on .gitignore:
```diff
@@ -1,82 +1,16 @@
-Debug
-build/
-
-## Eclipse IDE
```
.gitignore updated.

dongxuy04 commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,103 @@
+# This file is a template, and might need editing before it works on your project.
```
Thanks for pointing this out. It is not needed and deleted.
